modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/upgrade/v3_0_2/RedirectEntrySourceURLUpgradeProcess.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.redirect.internal.upgrade.v3_0_2;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Attila Bakay
+ */
+public class RedirectEntrySourceURLUpgradeProcess extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
+				"select redirectEntryId, sourceURL from RedirectEntry");
+			PreparedStatement preparedStatement2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update RedirectEntry set sourceURL = ? where " +
+						"redirectEntryId = ?")) {
+
+			ResultSet resultSet = preparedStatement1.executeQuery();
+
+			while (resultSet.next()) {
+				String sourceURL = resultSet.getString(2);
+				long redirectEntryId = resultSet.getLong(1);
+
+				_redirectEntries.put(redirectEntryId, sourceURL);
+			}
+
+			for (Map.Entry<Long, String> redirectEntry :
+					_redirectEntries.entrySet()) {
+
+				String sourceURL = redirectEntry.getValue();
+
+				String lowerCaseSourceURL = StringUtil.toLowerCase(sourceURL);
+
+				if (sourceURL.equals(lowerCaseSourceURL)) {
+					continue;
+				}
+
+				long redirectEntryId = redirectEntry.getKey();
+
+				if (!_hasRedirectEntry(redirectEntryId, lowerCaseSourceURL)) {
+					preparedStatement2.setString(1, lowerCaseSourceURL);
+					preparedStatement2.setLong(2, redirectEntryId);
+
+					preparedStatement2.addBatch();
+				}
+				else {
+					if (_log.isWarnEnabled()) {
+						_log.warn(
+							StringBundler.concat(
+								"Can not modify '", sourceURL, "' to '",
+								lowerCaseSourceURL,
+								"' because it is already in use id:  at id: ",
+								redirectEntryId));
+					}
+				}
+			}
+
+			preparedStatement2.executeBatch();
+		}
+	}
+
+	private boolean _hasRedirectEntry(
+		long redirectEntryId, String lowerCaseSourceURL) {
+
+		if (_duplicateSourceURLs.contains(lowerCaseSourceURL)) {
+			return true;
+		}
+
+		_duplicateSourceURLs.add(lowerCaseSourceURL);
+
+		for (Map.Entry<Long, String> redirectEntry :
+				_redirectEntries.entrySet()) {
+
+			String sourceURL = redirectEntry.getValue();
+
+			if ((redirectEntry.getKey() != redirectEntryId) &&
+				sourceURL.equals(lowerCaseSourceURL)) {
+
+				return true;
+			}
+		}
```

Comment:
This is very inefficient, and it looks redundant. All previously found urls will be already present in `_duplicateSourceURLs`, so there's no need to iterate over all elements again to see if it is duplicate. Unless I'm missing something, I think this new check it unnecessary.

Reply:
The _duplicateSourceURLs is populated based on the redirectEntries here. If a string is searched it will be added to the _duplicateSourceURLs but we have to check the Entries list for the first searches to avoid the collision. For exemple, if the entries like this:

id    SourceURLS
1       	/A
2      	/A
...	        ....

In this case, the first 'A' has a duplicate so we have to check the entries because the A is not in the _duplicateSourceURLs list yet.

Reply:
Oh, I see. To avoid the iteration process, why not adding an extra Set<String> that contains all seen sourceURLs? That way you could replace the loop with a simple `seenSourceURLs.contains(lowerCaseSourceURL)`. It is a memory/speed tradeoff, but will prevent the upgrade process to take a ton of time (n^2 vs logn).

Reply:
I'm not sure but I think that is what I do currently I'm only comparing every string( that is not part of the duplicate set) once to every other element and every element that is checked whether or not there were duplicates I add them to the duplicate list to speed up the search with the set contains. I'm doing the loop to avoid the collision with themselves so I could also check for the id.

Reply:
OR do you suggest creating a third collection just for the values to make a faster search?

Reply:
Nevermind, I think we can save that extra collection by mapping ids and urls reverse to what you're doing. If instead of Map<Long, String> you use a Map<String, Set<Long>> (url -> set of ids with that url) you could remove the loop, making that check constant time. Does that sound good to you? Thanks!

Reply:
I don't think that solution would work because of the unique index on the groupId SourceURL combo. But I have to modify the solution because it currently does not check if the sourceURL is unique in the current group. Currently, I only check if the sourceURL is unique in the DB.